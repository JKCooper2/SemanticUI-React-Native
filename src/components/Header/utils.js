// @flow
import _ from 'lodash';
import { fonts, colors } from '~/styles';

// TODO: Size should be rem vs as being em
const SIZE = {
    huge: 'h1',
    large: 'h2',
    medium: 'h3',
    small: 'h4',
    tiny: 'h5'
};

export const getFontSize = (asProp: string, size: string, fallback = 'h2'): number => {
    let asSize = size && SIZE[size]
        ? SIZE[size]
        : asProp;

    return _.get(fonts.size, [asSize], fonts.size[fallback]);
};

export const getFontColor = (color: string, inverted: boolean, disabled: boolean): string => {
    if (disabled) {
        return colors.fonts.disabled;
    }

    if (inverted) {
        _.get(colors.invertedColors, color, colors.fonts.heading);
    }

    return _.get(colors.colors, color, colors.fonts.heading);
};