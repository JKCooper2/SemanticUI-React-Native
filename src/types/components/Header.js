// @flow
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
import type { ColorsType, TextAlignType } from '../styles';

export type AsType = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
export type SizeType = "huge" | "large" | "medium" | "small" | "tiny";

export type HeaderPropsType = {
    as: AsType;
    size?: SizeType;
    sub: boolean;
    content: string;
    subheader: string;
    disabled: boolean;
    color: ColorsType;
    textAlign: TextAlignType;
    inverted: boolean;
    style?: StyleObj;
    children?: Node;
};

export type SubheaderPropsType = {
    children: Node,
    style?: StyleObj,
    as: AsType
};