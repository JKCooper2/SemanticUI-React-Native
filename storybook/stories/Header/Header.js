// @flow
import React, { Fragment } from 'react';
import { View, ScrollView } from 'react-native';

import { storiesOf } from '@storybook/react-native';

import Header from '~/components/Header';

const headerStories = storiesOf('Header', module);

headerStories.add('As', () => (
    <Fragment>
        <Header as="h1">Header H1</Header>
        <Header as="h2">Header H2</Header>
        <Header as="h3">Header H3</Header>
        <Header as="h4">Header H4</Header>
        <Header as="h5">Header H5</Header>
        <Header as="h6">Header H6</Header>
    </Fragment>
));

headerStories.add('Size', () => (
    <Fragment>
        <Header size="huge">Header Huge</Header>
        <Header size="large">Header Large</Header>
        <Header size="medium">Header Medium</Header>
        <Header size="small">Header Small</Header>
        <Header size="tiny">Header Tiny</Header>
    </Fragment>
));

headerStories.add('Sub', () => (
    <Header sub>SubHeader</Header>
));

headerStories.add('Content & Subheader', () => (
    <Header content="Header" subheader="Subheader" />
));

headerStories.add('Disabled', () => (
    <Header disabled>Header</Header>
));

headerStories.add('Text Align', () => (
    <View style={{ flex: 1 }}>
        <Header textAlign="left">Align Left</Header>
        <Header textAlign="center">Align Center</Header>
        <Header textAlign="right">Align Right</Header>
    </View>
));

headerStories.add('Color', () => (
    <ScrollView>
        <Header as='h4' color='red'>Red</Header>
        <Header as='h4' color='orange'>Orange</Header>
        <Header as='h4' color='yellow'>Yellow</Header>
        <Header as='h4' color='olive'>Olive</Header>
        <Header as='h4' color='green'>Green</Header>
        <Header as='h4' color='teal'>Teal</Header>
        <Header as='h4' color='blue'>Blue</Header>
        <Header as='h4' color='purple'>Purple</Header>
        <Header as='h4' color='violet'>Violet</Header>
        <Header as='h4' color='pink'>Pink</Header>
        <Header as='h4' color='brown'>Brown</Header>
        <Header as='h4' color='grey'>Grey</Header>
    </ScrollView>
));

headerStories.add('Inverted', () => (
    <ScrollView style={{ backgroundColor: 'black' }}>
        <Header as='h4' color='red' inverted>Red</Header>
        <Header as='h4' color='orange' inverted>Orange</Header>
        <Header as='h4' color='yellow' inverted>Yellow</Header>
        <Header as='h4' color='olive' inverted>Olive</Header>
        <Header as='h4' color='green' inverted>Green</Header>
        <Header as='h4' color='teal' inverted>Teal</Header>
        <Header as='h4' color='blue' inverted>Blue</Header>
        <Header as='h4' color='purple' inverted>Purple</Header>
        <Header as='h4' color='violet' inverted>Violet</Header>
        <Header as='h4' color='pink' inverted>Pink</Header>
        <Header as='h4' color='brown' inverted>Brown</Header>
        <Header as='h4' color='grey' inverted>Grey</Header>
    </ScrollView>
));

export default headerStories;