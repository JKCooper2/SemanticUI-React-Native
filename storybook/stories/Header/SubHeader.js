// @flow
import React, { Fragment } from 'react';

import { storiesOf } from '@storybook/react-native';

import Header from '~/components/Header';

const subheaderStories = storiesOf('SubHeader', module);

subheaderStories.add('Basic', () => <Header.SubHeader>Basic SubHeader</Header.SubHeader>);

subheaderStories.add('As', () => (
    <Fragment>
        <Header.SubHeader as="h1">SubHeader H1</Header.SubHeader>
        <Header.SubHeader as="h2">SubHeader H2</Header.SubHeader>
        <Header.SubHeader as="h3">SubHeader H3</Header.SubHeader>
        <Header.SubHeader as="h4">SubHeader H4</Header.SubHeader>
        <Header.SubHeader as="h5">SubHeader H5</Header.SubHeader>
        <Header.SubHeader as="h6">SubHeader H6</Header.SubHeader>
    </Fragment>
));

export default subheaderStories;
