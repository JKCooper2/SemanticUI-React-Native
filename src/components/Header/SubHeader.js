// @flow
import React, { PureComponent } from 'react';
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { Text } from 'react-native';
import _ from 'lodash';

import { getFontSize } from './utils';
import { colors } from '~/styles';

type Props = {
    children: Node,
    style?: StyleObj,
    as: "h1" | "h2" | "h3" | "h4" | "h5" | "h6",
};

class SubHeader extends PureComponent<Props> {
    static defaultProps = {
    };

    render() {
        const { children, style } = this.props;

        if (!_.isString(children)) {
            console.warn('A SubHeader must be passed a string child component');
            return children;
        }

        return (
            <Text style={[{ fontSize: getFontSize(this.props.as, '', 'h4'), color: colors.fonts.subheading }, style]}>{children}</Text>
        );
    }
}

export default SubHeader;