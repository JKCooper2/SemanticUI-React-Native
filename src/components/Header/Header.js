// @flow
import React, { PureComponent } from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';

import { spacing, colors } from '~/styles';

import type { HeaderPropsType } from '~/types/components/Header';

import SubHeader from './SubHeader';
import { getFontSize, getFontColor } from './utils';

// TODO: Icon, Image, Divider
// TODO: Block, Attached

class Header extends PureComponent<HeaderPropsType> {
    static SubHeader = SubHeader;

    static defaultProps = {
        sub: false,
        as: 'h2',
        disabled: false,
        content: '',
        subheader: '',
        color: colors.colors.black,
        textAlign: 'left',
        inverted: false
    };

    render() {
        const { children, size, sub, style, content, subheader, disabled, color, inverted, textAlign } = this.props;

        let textColor = getFontColor(color, inverted, disabled);

        let contentNode = null;

        if (content) {
            if (sub) {
                contentNode = <SubHeader style={style}>{content}</SubHeader>;
            } else {
                contentNode = <Text style={[{ fontSize: getFontSize(this.props.as, size), fontWeight: 'bold', color: textColor }, style]}>{content}</Text>;
            }
        }

        let childComp = React.Children.map(children, (child) => {
            if (!_.isString(child)) {
                return child;
            }

            return sub
                ? <SubHeader style={style}>{child}</SubHeader>
                : <Text style={[{ fontSize: getFontSize(this.props.as, size), fontWeight: 'bold', color: textColor }, style]}>{child}</Text>;
        });

        const alignSelf: {| alignSelf: string |} = _.get(spacing.textAlign, textAlign, spacing.textAlign.left);

        return (
            <View style={[{ padding: spacing.padding.large }, alignSelf]}>
                {contentNode}
                {childComp}
                {!!subheader && <SubHeader>{subheader}</SubHeader>}
            </View>
        );
    }
}

export default Header;
