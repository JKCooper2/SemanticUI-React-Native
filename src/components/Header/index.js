// @flow
import Header from './Header';
import SubHeader from './SubHeader';

export {
    Header as default,
    SubHeader
};