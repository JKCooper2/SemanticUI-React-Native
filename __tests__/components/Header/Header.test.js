// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Header from '~/components/Header';

test('renders correctly', () => {
    const header = shallow(<Header />);
    expect(header).toMatchSnapshot();
});
