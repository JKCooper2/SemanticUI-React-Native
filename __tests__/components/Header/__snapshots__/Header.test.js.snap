// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Array [
      Object {
        "padding": 30,
      },
      Object {
        "alignSelf": "flex-start",
      },
    ]
  }
/>
`;
